aiida/cmdline/commands/cmd_process.py

```diff
@@ -347,6 +348,8 @@ def process_actions(futures_map, infinitive, present, past, wait=False, timeout=
             process = futures_map[future]
 
             try:
+                # unwrap is need here since LoopCommunicator will also wrap a future
+                future = unwrap_kiwi_future(future)
```

Comment:
You are right that this part is not very clear. And I think there is the same issue for the `aiida_core` of the tornado version. 

Let me try to describe my understanding of this part of the code:
The future returned from the `controller.kill_process` is not  depend on the communicator we used here, but on communicator which used to actually doing the `kill` action. That is the `LoopCommunicator` and we add a rpc_subscriber in that communicator of the daemon `Runner`. 
The reason that we do not need use the LoopCommunicator in `verdi process` is every time we run a cmd command we run it in a independent thread and wait for the returned value. 
And it is because we don't know what communicator we used exactly, we need to put a generic `unrap_kiwi_future` here to get the final result of the procedure scheduled in the communicator.

Reply:
Haha, yes, I'm also getting confused looking through this part of the code.  I guess the principle that Sebastiaan is pointing out (and I agree with) is that if there is a `Communicator` interface that we're using then things that come out of the interface should always be at the same level of 'unwrapping', ideally such that no manual unwrapping needs to be done outside, and certainly in a way that doesn't vary on the underlying concrete `Communicator` instance type.  

I'll have a look through the code and see if I can make sure I understand what the flow is.